src/main/java/org/logstash/beats/BeatsParser.java

```diff
@@ -0,0 +1,230 @@
+        if(!hasEnoughBytes(in)) {
+            return;
+        }
+
+        switch (currentState) {
+            case READ_HEADER: {
+                logger.debug("Running: READ_HEADER");
+
+                byte currentVersion = in.readByte();
+
+                if(Protocol.isVersion2(currentVersion)) {
+                    logger.debug("Frame version 2 detected");
+                    batch.setProtocol(Protocol.VERSION_2);
+                } else {
+                    logger.debug("Frame version 1 detected");
+                    batch.setProtocol(Protocol.VERSION_1);
+                }
+
+                transition(States.READ_FRAME_TYPE, 1);
+                break;
+            }
+            case READ_FRAME_TYPE: {
+                byte frameType = in.readByte();
+
+                switch(frameType) {
+                    case Protocol.CODE_WINDOW_SIZE: {
+                        transition(States.READ_WINDOW_SIZE, 4);
+                        break;
+                    }
+                    case Protocol.CODE_JSON_FRAME: {
+                        // Reading Sequence + size of the payload
+                        transition(States.READ_JSON_HEADER, 8);
+                        break;
+                    }
+                    case Protocol.CODE_COMPRESSED_FRAME: {
+                        transition(States.READ_COMPRESSED_FRAME_HEADER, 4);
+                        break;
+                    }
+                    case Protocol.CODE_FRAME: {
+                        transition(States.READ_DATA_FIELDS, 8);
+                        break;
+                    }
+                }
+                break;
+            }
+            case READ_WINDOW_SIZE: {
+                logger.debug("Running: READ_WINDOW_SIZE");
+
+                batch.setBatchSize((int) in.readUnsignedInt());
+
+                // This is unlikely to happen but I have no way to known when a frame is
+                // actually completely done other than checking the windows and the sequence number,
+                // If the FSM read a new window and I have still
+                // events buffered I should send the current batch down to the next handler.
+                if(!batch.isEmpty()) {
+                    logger.warn("New window size received but the current batch was not complete, sending the current batch");
+                    out.add(batch);
+                    batchComplete();
+                }
+
+                transitionToReadHeader();
+                break;
+            }
+            case READ_DATA_FIELDS: {
+                // Lumberjack version 1 protocol, which use the Key:Value format.
+                logger.debug("Running: READ_DATA_FIELDS");
+                sequence = (int) in.readUnsignedInt();
+                int fieldsCount = (int) in.readUnsignedInt();
+                int count = 0;
+
+                Map dataMap = new HashMap<String, String>();
```

Comment:
you can size the hashmap appropriately by putting `fieldsCount` as constructor argument